### PR TITLE
Handle specific exception codes on RCI call.

### PR DESCRIPTION
**Pull request by BinSquare**
### Summary

1. This adds a terminalError wrapper around 2 exception codes that are known to not be recoverable by retrying (this can help avoid retry storm). This data was verified from internal investigation. 
2. Separately, it also checks that throttlingError does return exitcode 1 i.e. it will be retried. 
3. Refactored from isTransient to isTerminal because the terminal cases are what we want to specifically handle while the default is to delegate to existing retry mechanism for other cases.

### Implementation details

Changes are implemented on agent calling RCi.

### Testing

Added 2 new tests, one to verify throttlingError returns exitcode 1 + one to verify ClientException will result in exit code 5 i.e. the terminal state for agent.

New tests cover the changes: 

### Description for the changelog

Terminally exit on unrecoverable RCI exception errors.

### Additional Information

**Does this PR include breaking model changes? If so, Have you added transformation functions?**
  
No

**Does this PR include the addition of new environment variables in the README?**

No

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/app/errors.go [hunk 1]
```diff
@@ -24,6 +24,18 @@ func isTransient(err error) bool {
...
 
+type terminalError struct {
+	error
+}
+
+// isTerminal returns true if the error is already wrapped as an unrecoverable condition
+// which will allow agent to exit terminally.
+func isTerminal(err error) bool {
+	// Check if the error is already wrapped as a terminalError
+	_, terminal := err.(terminalError)
```
Comment (timj-hh) [inline agent/app/errors.go:35]: Could there be a situation in which this gets wrapped again? Why not use a deeper equality check like [errors.Is](https://pkg.go.dev/errors#Is) which inspects the error tree rather than its first layer?

File: agent/app/agent.go [hunk 2]
```diff
@@ -447,10 +447,15 @@ func (agent *ecsAgent) doStart(containerChangeEventStream *eventstream.EventStre
 	// Register the container instance
 	err = agent.registerContainerInstance(client, vpcSubnetAttributes)
 	if err != nil {
-		if isTransient(err) {
+		if isTerminal(err) {
+			// On unrecoverable error codes, agent should terminally exit.
+			seelog.Warnf("Agent will terminally exit, unable to register container instance: %v", err)
```
Comment (timj-hh) [inline agent/app/agent.go:452]: nit: should use `logger` instead of seelog here?

**Review by timj-hh — APPROVED** [review]

File: agent/app/agent.go (hunk 2 above)
Comment (BinSquare) [reply agent/app/agent.go:452]: seelog.Warnf is chosen here to conform to existing style of using it for warning messages in this specific file i.e. there were 9 seelog.warn vs 2 logger.warn - so I opted for the more commonly used

There are 9 cases I used as reference found :
1. https://github.com/aws/amazon-ecs-agent/blob/41d593c6110b7dc43bea3db6f871a15591917af0/agent/app/agent.go#L945
2. https://github.com/aws/amazon-ecs-agent/blob/41d593c6110b7dc43bea3db6f871a15591917af0/agent/app/agent.go#L953


Where as logger is also used but mostly for logger.critical and logger.info. There are only 2 cases of logger.Warn: https://github.com/aws/amazon-ecs-agent/blob/41d593c6110b7dc43bea3db6f871a15591917af0/agent/app/agent.go#L1036

File: agent/app/errors.go (hunk 1 above)
Comment (BinSquare) [reply agent/app/errors.go:35]: There should not be a case where it gets wrapped again - the terminal wrap should be the last layer as it's handled in this file. So in this case, both errors.Is and the current behaves the same.

Since existing isTransient is implemented similarly - hopefully this optimization shouldn't be a blocker?

File: agent/app/agent.go [hunk 3]
```diff
@@ -840,7 +845,13 @@ func (agent *ecsAgent) registerContainerInstance(
...
-			return err
+			return terminalError{err}
+		}
+		if utils.IsAWSErrorCodeEqual(err, ecsmodel.ErrCodeClientException) {
+			logger.Critical("Instance registration attempt with client performing invalid action", logger.Fields{
+				field.Error: err,
+			})
+			return terminalError{err}
 		}
 		if _, ok := err.(apierrors.AttributeError); ok {
```
Comment (xxx0624) [inline agent/app/agent.go:856]: Is it possible this is also terminal error? 
> Instance registration attempt with invalid attribute(s)

**Review by xxx0624 — APPROVED** [review]

File: agent/app/agent.go (hunk 3 above)
Comment (BinSquare) [reply agent/app/agent.go:856]: Handling that specific case is beyond scope of this change.

However, invalidParameterException is handled above, I would not expect attributeError to fire because instance attributes are an instance parameter of the RCI call. Ecs agent should correct me here if I'm wrong

File: agent/app/errors.go (hunk 1 above)
Comment (timj-hh) [reply agent/app/errors.go:35]: This is not a blocker, but in the case where we re-use these errors it may be a better long term change to wrap/unwrap these with the errors library.

File: agent/app/agent.go (hunk 2 above)
Comment (timj-hh) [reply agent/app/agent.go:452]: Got it, thanks for the justification. I am no expert but it seems more idiomatic to conform to the in-house logger used in this library

File: agent/app/agent.go (hunk 3 above)
Comment (xxx0624) [reply agent/app/agent.go:856]: Sounds good to me. We can focus on the cases we care about now.

File: agent/app/agent.go [hunk 4]
```diff
@@ -447,10 +447,15 @@ func (agent *ecsAgent) doStart(containerChangeEventStream *eventstream.EventStre
...
+		if isTerminal(err) {
+			// On unrecoverable error codes, agent should terminally exit.
+			seelog.Warnf("Agent will terminally exit, unable to register container instance: %v", err)
+			return exitcodes.ExitTerminal
+		} else if isTransient(err) {
 			return exitcodes.ExitError
 		}
-		return exitcodes.ExitTerminal
+		// Other errors are considered recoverable and will be retried.
+		return exitcodes.ExitError
```
Comment (amogh09) [inline agent/app/agent.go:458]: Why do we want to change the default?

Comment (singholt) [reply agent/app/agent.go:458]: +1

File: agent/app/agent.go (hunk 2 above)
Comment (singholt) [reply agent/app/agent.go:452]: also, I think a log statement for a terminal error should be logged with "Critical" level instead of a "Warn".

Comment (BinSquare) [reply agent/app/agent.go:452]: Ack - implementing recommended

File: agent/app/agent.go (hunk 4 above)
Comment (BinSquare) [reply agent/app/agent.go:458]: Pasting from convo: A terminal exit for the agent is a last resort case because it's painful in terms of task experience - so we want to be very specific about when to terminally exit rather than have it be the catch-all default.

**Comment by BinSquare** [comment]
@amogh09 has a separate PR which will address the static check failure due to the copyright date change from 2024 -> 2025 here: https://github.com/aws/amazon-ecs-agent/pull/4461

File: agent/app/agent_test.go [hunk 5]
```diff
@@ -1499,6 +1499,131 @@ func TestRegisterContainerInstanceInvalidParameterTerminalError(t *testing.T) {
 		credentialsManager, state, imageManager, client, execCmdMgr)
 	assert.Equal(t, exitcodes.ExitTerminal, exitCode)
 }
+
+func TestRegisterContainerInstanceInvalidClientError(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (BinSquare) [reply agent/app/agent_test.go]: Implemented the suggested

**Review by singholt — DISMISSED** [review]

File: agent/app/agent_test.go [hunk 6]
```diff
@@ -795,7 +795,7 @@ func TestNewTaskEngineRestoreFromCheckpointNewStateManagerError(t *testing.T) {
 	_, _, err := agent.newTaskEngine(eventstream.NewEventStream("events", ctx),
 		credentialsManager, dockerstate.NewTaskEngineState(), imageManager, hostResources, execCmdMgr, serviceConnectManager, daemonManagers)
 	assert.Error(t, err)
-	assert.False(t, isTransient(err))
+	assert.False(t, isTerminal(err))
```
Comment (amogh09) [inline agent/app/agent_test.go:798]: This new assertion is a bit odd (it is odd before the change too). `newTaskEngine` method does interact with `transientError` or `terminalError` wrapper types at all. It does not modify the error returned to it by `stateManagerFactory.NewStateManager` method. It is called in `doStart` method which treats all `newTaskEngine` failures as terminal failures. 

I think we should simply assert that the error returned by `newTaskEngine` is the same as that returned by `stateManagerFactory.NewStateManager`. 

```go
assert.EqualError(t, err, "error")
```

or if you want to be more strict 
```go
someError := errors.New("error")
...
    stateManagerFactory.EXPECT().NewStateManager(gomock.Any(), gomock.Any(),
			gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(nil, someError),
...
assert.Error(t, err)
assert.Equal(t, err, someError)
```

File: agent/app/agent_test.go [hunk 7]
```diff
@@ -844,7 +844,7 @@ func TestNewTaskEngineRestoreFromCheckpointStateLoadError(t *testing.T) {
...
```
Comment (amogh09) [inline agent/app/agent_test.go:847]: Same comment as above.

File: agent/app/agent_test.go (hunk 6 above)
Comment (BinSquare) [reply agent/app/agent_test.go]: Agreed that it's weird (and not useful) for both before/after - but a deep dive finds that it's not useful to implement the error check either. The error returned by NewStateManager is wrapped by newTaskEngine here: https://github.com/aws/amazon-ecs-agent/blob/41d593c6110b7dc43bea3db6f871a15591917af0/agent/app/agent.go#L597

Ex.
```
        	Error Trace:	/home/runner/work/amazon-ecs-agent/amazon-ecs-agent/src/github.com/aws/amazon-ecs-agent/agent/app/agent_test.go:799
        	Error:      	Not equal: 
        	            	expected: *errors.withStack(failed to load previous data from state file: error)
```

meaning we will wrap the injected error with "failed to load previous data...". 

---

Since the code path touched doesn't have specific treatment for errors beyond cases already touched upon by new test cases - it's not particularly useful in this case to introduce injecting an error and checking the wrapped injected error. 

I think it's sufficient to just ensure the function errors to keep in line with spirit of 
```
func TestNewTaskEngineRestoreFromCheckpointStateLoadError(t *testing.T) {
```

File: agent/app/agent_test.go (hunk 7 above)
Comment (BinSquare) [reply agent/app/agent_test.go]: Same comment as above:

Quick summary is that it’s not useful to implement a specific error check here. The tested codepath doesn’t handle different errors in a meaningful way as opposed to the added cases in RCI. As such it's sufficient to check the error as we previously did and the new edge cases are covered by the new test cases.

File: agent/app/agent_test.go (hunk 6 above)
Comment (amogh09) [reply agent/app/agent_test.go]: `seelog.Criticalf` does not create or change an error. 

The wrapping actually happens here 
https://github.com/aws/amazon-ecs-agent/blob/41d593c6110b7dc43bea3db6f871a15591917af0/agent/app/data.go#L98

So a tighter assertion would be - 
```
assert.EqualError(t, err, "failed to load previous data from state file: error")
```

But I agree that might be outside the scope of this PR and can be done separately.

**Review by singholt — APPROVED** [review]
